Bluetooth: controller: Check preempt event on timeout

cvinayak opened the pull request:
Check whether the preempt event matches with the head of the
pipeline before aborting the currently active event.

This is required to avoid preemption of events that became
active due to done event and there has been a race in
stopping the preempt ticker.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: openisa: Preempt timeout for next in pipeline

Start new preempt timeout for the next in pipeline when a
new event gets active.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: nRF5: Preempt timeout for next in pipeline

Start new preempt timeout for the next in pipeline when a
new event gets active.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: nRF5: Check preempt event on timeout

Check whether the preempt event matches with the head of the
pipeline before aborting the currently active event.

This is required to avoid preemption of events that became
active due to done event and there has been a race in
stopping the preempt ticker.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: openisa: Check preempt event on timeout

Check whether the preempt event matches with the head of the
pipeline before aborting the currently active event.

This is required to avoid preemption of events that became
active due to done event and there has been a race in
stopping the preempt ticker.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot reviewed (approved):
I like it. It plays very well into the new scheduling concept, and the changes needed for that.

cvinayak committed:
Bluetooth: controller: openisa: Sync up LLL changes

Sync up with missing LLL changes done related to moving the
handling of pipeline handling into LLL.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented:
@thoh-ot and @mtpr-ot could you please review so we can include this in 2.6?

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -785,43 +816,63 @@ static void preempt(void *param)
-	ret = event.curr.is_abort_cb(next->prepare_param.param,
+	/* Preemptor not in pipeline */
+	if (next != param) {
+		uint32_t ret;
+
+		/* Start the preempt timeout */
+		ret = preempt_ticker_start(next, ticker_start_next_op_cb);
+		LL_ASSERT((ret == TICKER_STATUS_SUCCESS) ||
+			  (ret == TICKER_STATUS_BUSY));
```
@mtpr-ot This is the change compare to your previous review. When a event in pipeline gets active before preempt timeout, then next event in pipeline shall setup new timeout. Otherwise, in the case of continuous scanning the window is not closed and pipeline overflows after a number of scan intervals as there is nothing to close the window, and done event to dequeue  the pipeline.

mtpr-ot reviewed (approved):
LGTM
